CenterForOpenScience/ember-osf-web
# [ENG-1383] Draft reg manager class

- Ticket: [ENG-1383]
- Feature flag: n/a

## Purpose
- Add new child routes to the `drafts/draft` route
  - metadata and review routes added
  - Move route specific logic (page route, `inReview`, etc) to appropriate places
- No longer use DraftRegistrationManager as a component, but use it as a class instead
  - Needs to be a class on the `drafts/draft` route level to be accessed by all of the child routes 
  - Divide old DRM responsibilities to other classes/routes/etc

## Summary of Changes
- Create new DraftRegManager class (handle validations, saving, instantiating pageManagers)
- Create new NavigationManager class to handle navigation between routes and pages (pageManager, pageParams, pageIndex)
  - Update TopNav, LeftNav > PageLink and RightNav  to use this class
- Move layout logic from `drafts/draft/page` to `drafts/draft` template and have child routes handle content specific to each route
- set queryParams `refreshModel` to `false` (needed to prevent the `draft` model hook from running every time route changes)
  - Tests that check the URL were updated
  - New utility (window-location) is used for some view-only-link tests that reloads the app (a stub is used in tests)
- Update the way helpText is mapped for single- and multi-select
  

## Side Effects
- Yes (more info coming soon)

## QA Notes
Since this is a PR that will fundamentally change how the draft registration is handled, another test through of the draft registration workflow would be greatly appreciated :)
- What pages should be tested?
  - The draft registration workflow should be double checked to make sure all behavior (navigation, saving inputs, displaying labels, help texts, etc) are working as expected. 
- Is cross-browser testing required/recommended?
  - Cross-browser and mobile testing would be _mucho_ recommended.
- What edge cases should QA be aware of?
  - None that I am aware of, but I'm sure there's plenty that I haven't considered
- What level of risk would you expect these changes to have?
  - DEFCON 1. Code red. 1 minute to midnight.
- For each feature flag (if any), what is the expected behavior with the flag enabled vs disabled?
  - N/A


[ENG-610]: https://openscience.atlassian.net/browse/ENG-610

[ENG-1383]: https://openscience.atlassian.net/browse/ENG-1383
